lightning/src/ln/features.rs

```diff
@@ -725,26 +725,40 @@ macro_rules! impl_feature_tlv_write {
 	($features: ident) => {
 		impl Writeable for $features {
 			fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
-				self.write_be(w)
+				WithoutLength(self).write(w)
 			}
 		}
 		impl Readable for $features {
 			fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
-				let v = io_extras::read_to_end(r)?;
-				Ok(Self::from_be_bytes(v))
+				Ok(WithoutLength::<Self>::read(r)?.0)
 			}
 		}
 	}
 }
 
 impl_feature_tlv_write!(ChannelTypeFeatures);
-impl_feature_tlv_write!(OfferFeatures);
-impl_feature_tlv_write!(InvoiceRequestFeatures);
```

Comment:
Tests are passing for me on #1926 with keeping use of the macro for these sets of features:
```
diff --git a/lightning/src/ln/features.rs b/lightning/src/ln/features.rs
index a3c1b7f6..335dad03 100644
--- a/lightning/src/ln/features.rs
+++ b/lightning/src/ln/features.rs
@@ -751,6 +751,8 @@ macro_rules! impl_feature_tlv_write {
 }

 impl_feature_tlv_write!(ChannelTypeFeatures);
+impl_feature_tlv_write!(OfferFeatures);
+impl_feature_tlv_write!(InvoiceRequestFeatures);

 // Some features may appear both in a TLV record and as part of a TLV subtype sequence. The latter
 // requires a length but the former does not.
@@ -872,11 +874,11 @@ mod tests {
                assert_eq!(features.flags.len(), 8);

                let mut serialized_features = Vec::new();
-               WithoutLength(&features).write(&mut serialized_features).unwrap();
+               features.write(&mut serialized_features).unwrap();
                assert_eq!(serialized_features.len(), 8);

                let deserialized_features =
-                       WithoutLength::<OfferFeatures>::read(&mut &serialized_features[..]).unwrap().0;
+                       OfferFeatures::read(&mut &serialized_features[..]).unwrap();
                assert_eq!(features, deserialized_features);
        }

diff --git a/lightning/src/offers/invoice_request.rs b/lightning/src/offers/invoice_request.rs
index fc2b90f8..46f7b626 100644
--- a/lightning/src/offers/invoice_request.rs
+++ b/lightning/src/offers/invoice_request.rs
@@ -408,7 +408,7 @@ impl Writeable for InvoiceRequestContents {
 tlv_stream!(InvoiceRequestTlvStream, InvoiceRequestTlvStreamRef, 80..160, {
        (80, chain: ChainHash),
        (82, amount: (u64, HighZeroBytesDroppedBigSize)),
-       (84, features: (InvoiceRequestFeatures, WithoutLength)),
+       (84, features: InvoiceRequestFeatures),
        (86, quantity: (u64, HighZeroBytesDroppedBigSize)),
        (88, payer_id: PublicKey),
        (89, payer_note: (String, WithoutLength)),
diff --git a/lightning/src/offers/offer.rs b/lightning/src/offers/offer.rs
index d92d0d8b..04acc4da 100644
--- a/lightning/src/offers/offer.rs
+++ b/lightning/src/offers/offer.rs
@@ -578,7 +578,7 @@ tlv_stream!(OfferTlvStream, OfferTlvStreamRef, 1..80, {
        (6, currency: CurrencyCode),
        (8, amount: (u64, HighZeroBytesDroppedBigSize)),
        (10, description: (String, WithoutLength)),
-       (12, features: (OfferFeatures, WithoutLength)),
+       (12, features: OfferFeatures),
        (14, absolute_expiry: (u64, HighZeroBytesDroppedBigSize)),
        (16, paths: (Vec<BlindedPath>, WithoutLength)),
        (18, issuer: (String, WithoutLength)),
```

Reply:
That is expected because `impl_feature_tlv_write` should be equivalent. I suppose ideally we get rid of the macro and make it explicit for `ChannelTypeFeatures`.

Reply:
I'm confused why get rid of the macro, it seems cleaner to have it and avoid `WithoutLength` in the offers code 🤔

Reply:
Seems confusing if `Features` encoding were determined by the type of features rather than the context in which they are encoded. While features for the BOLT 12 message are currently only encoded inside a TLV record, that may not always be the case. Better to have the context be explicit about the encoding than to decide now on how the features must be encoded in all contexts, even unknown future contexts.

Reply:
Aren't all features encoded the same everywhere except this weird blinded payinfo one? Seems weird to make several callsites uglier because of one exception, or at least better to do them all on one go rather than have a few flexible exceptions for offers-related feature sets only. Don't think it's weird enough to hold up the PR but just trying to clarify.

Reply:
Oh, I just meant that for any given feature set X with the exception of blinded_payinfo_features, X will always be encoded the same way. 

>  But that's not to say that any of the other feature types won't be used in a similar scenario in the future.

I just don't see what that scenario would be 🤔 this seems like a one-off weird one.

Reply:
At very least, if there is a such a scenario, the compiler would complain. Whereas it won't complain when defining the serialization with `impl_feature_tlv_write`.